pkg/audit/manager.go

```diff
@@ -0,0 +1,270 @@
+		rs, err := clientset.Discovery().ServerResourcesForGroupVersion(constraintsGV)
+		if err != nil {
+			return err
+		}
+		// get constraints for each kind
+		for _, r := range rs.APIResources {
+			log.Info("constraint", "resource kind", r.Kind)
+			resourceGvk := strings.Split(rs.GroupVersion, "/")
+			group := resourceGvk[0]
+			version := resourceGvk[1]
+			resourceClient := dynamic.Resource(schema.GroupVersionResource{Group: group, Version: version, Resource: r.Name})
+			l, err := resourceClient.List(metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			// get and clear/reset each constraint violation if it's not currently empty
+			for _, item := range l.Items {
+				oname := item.GetName()
+				oapiVersion := item.GetAPIVersion()
+				ogvk := item.GroupVersionKind()
+				okind := ogvk.Kind
+				ogroup := ogvk.Group
+				oversion := ogvk.Version
+				violations, found, err := unstructured.NestedSlice(item.Object, "status", "violations")
+				if err != nil {
+					return err
+				}
+				if !found || len(violations) == 0 {
+					fmt.Printf("constraint %s violations is either not found or empty, skip clear \n", oname)
+				} else {
+					fmt.Printf("constraint name %s, apiversion %s, kind %s, group %s, version %s, violations %v \n", oname, oapiVersion, okind, ogroup, oversion, violations)
+					var emptyAuditResults []auditResult
+					err = updateConstraintStatus(dynamic, ogroup, oversion, okind, oname, oapiVersion, emptyAuditResults)
+					if err != nil {
+						return err
+					}
+					fmt.Printf("constraint %s violations have been cleared \n", oname)
+				}
+			}
+		}
+		// update each violated constraint
+		for selfLink := range updateList {
+			auditResults := updateList[selfLink]
+			aResult := auditResults[0]
+			return updateConstraintStatus(dynamic, aResult.cgroup, aResult.cversion, aResult.ckind, aResult.cname, aResult.capiversion, auditResults)
+		}
+	}
+	return nil
+}
+
+func (am *auditManager) auditManagerLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Audit Manager close")
+			close(am.stopper)
+			return
+		default:
+			time.Sleep(time.Duration(*auditInterval) * time.Second)
+			if err := am.audit(); err != nil {
+				log.Error(err, "audit manager audit() failed")
+			}
+		}
+	}
+}
+
+func startMgr(mgr manager.Manager, stopper chan struct{}, stopped chan<- struct{}) {
+	if err := mgr.Start(stopper); err != nil {
+		log.Error(err, "Error starting audit watch manager")
+	}
+	// mgr.Start() only returns after the manager has completely stopped
+	close(stopped)
+	log.Info("audit exiting")
+}
+
+func updateConstraintStatus(dynamic dynamic.Interface, cgroup string, cversion string, ckind string, cname string, capiversion string, auditResults []auditResult) error {
+	cresource := fmt.Sprintf("%ss", strings.ToLower(ckind))
+	cstrClient := dynamic.Resource(schema.GroupVersionResource{Group: cgroup, Version: cversion, Resource: cresource})
```

Comment:
I'm also curious if we can get away with using the regular client here and other places where we are GET-ing constraints. For some reason it works with updating the status and finalizers on constraints in the constraint controller. It might make life easier.

The regular client does fail in the unit test code for ConstraintTemplate. I'm unsure what's different. A preceding get?